org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/syntaxcoloring/ISemanticHighlightingIdMapper.xtend

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.editor.syntaxcoloring
+
+import com.google.common.base.Preconditions
+import com.google.common.base.Suppliers
+import com.google.inject.ImplementedBy
+import com.google.inject.Inject
+import com.google.inject.Provider
+import com.google.inject.Singleton
+import java.util.List
+import java.util.function.Supplier
+import org.eclipse.xtext.LanguageInfo
+
+/**
+ * Service for mapping the IDs of the highlighting styles to <a href="https://manual.macromates.com/en/language_grammars">TextMate scopes</a>.
+ * 
+ * <p>
+ * Clients can safely ignore this service and stick to the {@link ISemanticHighlightingIdMapper.Default default implementation}
+ * as is if they are not interested in the language server support. Otherwise, they have to map all styles used by the {@link ISemanticHighlightingCalculator}
+ * to the corresponding TextMate scopes.
+ * 
+ * @see IHighlightedPositionAcceptor#addPosition(int, int, String...)
+ * @see LightweightPosition#getIds()
+ *
+ * @since 2.15
+ */
+@ImplementedBy(Default)
+interface ISemanticHighlightingIdMapper {
+
+	/**
+	 * Maps the highlighting style ID to the corresponding TextMate scopes.
+	 */
+	def List<String> map(String id);
+
+	/**
+	 * The default ID mapper. Does nothing at all but appends the {@link LanguageInfo#getShortName() short name}
+	 * of the language.
+	 * 
+	 * <p>
+	 * For instance, if the ID of the highlighting style is {@code foo} and the name of your language
+	 * is {@code org.your.language.Bar}, then the resulting ID will be {@code foo.Bar.xtext}.
```

Comment:
is the shortname sufficient enhough? or do you assume its unique?

Reply:
No, ID uniqueness is not a must.

If a language maps a style string to `['foo', 'bar']` and another language maps a different style string to `['foo', 'bar']` too, then they [will be merged together](https://github.com/eclipse/xtext-core/pull/815/files#diff-4c70201045ee55f33ea5a3df3a6944a1R88) by the highlighting manager.

Here is the corresponding test case: https://github.com/eclipse/xtext-core/pull/815/files#diff-efe3ee25966b9c9a9b9ce3b5f496274eR47

If it needs a bit more explanation, please let me know.

Reply:
why is the shortname then needed at all?

Reply:
It is not needed, but it is still easier to debug and track the style-TextMate scope mappings.